limited base filename to 230 characters to be able to download audiobooks with long filenames. added filename-mode option: asin_only

devvythelopper opened the pull request:
this is a fix for https://github.com/mkb79/audible-cli/issues/215

mkb79 commented on src/audible_cli/models.py:
```diff
@@ -95,13 +95,18 @@ def create_base_filename(self, mode: str):
             base_filename = self.asin + "_" + base_filename
 
-        return base_filename
+        encoded = base_filename.encode('utf-8')
```
Why do you have encoded the filename to bytes before you slice?

devvythelopper replied:
because most filesystems are limited to byte numbers, not to characters. UTF-8 characters might need up to 4 bytes.

devvythelopper replied:
and luckily this code works in case of ascii as well, because ascii is a subset of utf-8

devvythelopper replied:
and this is the reason why your suffixes (`-annotations.json`) are safe, because you know their maximum length, and you know they are ascii.

mkb79 replied:
> because most filesystems are limited to byte numbers, not to characters. UTF-8 characters might need up to 4 bytes (think chinese characters).

I does not was aware about this behavior. Thank you for this "lesson". 
On `ascii` or `asin_ascii` mode, the filename is get by `BaseItem.full_title_slugify` property. So there are only `ASCII` letters. But I will re-add the byte encoding now.

devvythelopper replied:
what I meant by "safe": when you use UTF-8 and you have a chinese book name and you cut that off at 230 bytes. And then you add your suffix (`annotations.json`) and since your suffix **consists of** only ASCII character, each character will only add 1 byte. And so including your suffix the complete filename will be guaranteed to be less than 254 bytes.

crueber commented:
I ended up using @devvythelopper 's github repo's copy because this PR isn't merged, and I couldn't pull several very long filenames still. You should consider merging for the `asin_only` file mode.

devvythelopper commented on src/audible_cli/models.py:
```diff
@@ -82,26 +82,31 @@ def full_title_slugify(self):
+
+        # limiting 230 bytes, so that a suffix (e.g. -annotations.json) can be added easily
+        limited_bytes = encoded[:max_length]
+        return limited_bytes.decode("utf-8", errors="ignore")
```
No it is not possible to implement a "more robust truncation", as the only three options are to 

1. drop a multi-byte character whose trailing bytes are missing, and create a basename that is less than 230 (minimum 227) bytes long (errors="ignore")
2. replace a multi-byte character whose trailing bytes are missing with a placeholder and thus also have a basename that is less than 230 bytes (minimum 228) long
3. detect where the last character ends, and thus create a basename that is more than 230 bytes (maximum 233 bytes) long

All three solutions cut off either before the last or after the last character within the 230 bytes. Number 3 is needless work, and does not mitigate any problem (there will still be a cut off after the last character within those 230 bytes. Number 2 introduces a placeholder character that is ugly. Hence Number 1 is the most desirable, leave it as it is.

crueber commented:
It's true, I ended up not using the asin_only. So perhaps I was wrong there- Unfortunately, I ended up having to roll my own solution for one book that was just so darn long that even once I updated to the latest version of the github repo that each download still tried to push data to the same file. I ended up downloading the files by hand from my Audible library, and then had to write a special script to transcode them, because the metadata caused issues with the normal `AAXtoMP3` process.

mkb79 commented:
I'll on a trip until tomorrow. Will take a look this weekend and, if all is fine, I will implement it.

devvythelopper commented:
> It's true, I ended up not using the asin_only after that was the only way I found to download it, because the transcode wouldn't work either. Unfortunately, I ended up having to roll my own solution for that one book, because the metadata on transcode would cause major issues too. I ended up downloading the files by hand from my Audible library, and then had to write a special script to transcode them, because the metadata caused issues with the normal `AAXtoMP3` process.

@crueber My guess is you likely stumbled upon the issue, that audible-cli by default outputs the chapters in the chapters.json as a hierarchy (sub-chapters within chapters). However aaxtomp3 cannot deal with this format. To get a format that aaxtomp3 can deal with, you can specify the argument `--chapter-type Flat` which @mkb79 introduced not too long ago into the master branch. The audible-cli version in my repo already has it.

mkb79 commented:
@devvythelopper @crueber 
I've implemented an `--filename-length` option to the download command and changed error handling when decoding bytes to str to use replace. Maybe you can try out this pr and report back if this ist running for you?!

devvythelopper commented on src/audible_cli/models.py:
```diff
@@ -82,26 +82,31 @@ def full_title_slugify(self):
+
+        # limiting 230 bytes, so that a suffix (e.g. -annotations.json) can be added easily
+        limited_bytes = encoded[:max_length]
+        return limited_bytes.decode("utf-8", errors="replace")
```
Now Copilot suggests the opposite of its previous suggestion.

devvythelopper commented on src/audible_cli/models.py:
```diff
@@ -82,26 +82,31 @@ def full_title_slugify(self):
-
         else:
-            base_filename = self.asin
+            raise AudibleCliException(f"Unsupported mode {mode} for name creation")
```
I already commited the raise before. The only better option would be to rewrite the logic.

devvythelopper commented:
I already commented on Copilots original suggestion to change to errors="replace":

- drop a multi-byte character whose trailing bytes are missing, and create a basename that is less than 230 (minimum 227) bytes long (errors="ignore")
- replace a multi-byte character whose trailing bytes are missing with a placeholder and thus also have a basename that is less than 230 bytes (minimum 228) long (errors="replace")
- detect where the last character ends, and thus create a basename that is more than 230 bytes (maximum 233 bytes) long (which could not be called a "more robust truncation" either).

All three solutions cut off either before the last or after the last character within the 230 bytes. Number 3 is needless work, and does not mitigate any problem (there will still be a cut off after the last character that began within those 230 bytes). Number 2 introduces a placeholder character that is ugly.

So Copilot's original suggestion should simply be ignored, and `errors="ignore"` is the most desirable of the options.

devvythelopper commented:
> @devvythelopper @crueber I've implemented an `--filename-length` option to the download command and changed error handling when decoding bytes to str to use replace. Maybe you can try out this pr and report back if this ist running for you?!

I am getting the error `download_aaxc() missing 1 required positional argument: 'filename_length'` despite having specified `--filename-length 200` on the command line.

devvythelopper commented on src/audible_cli/models.py:
```diff
@@ -82,26 +82,31 @@ def full_title_slugify(self):
+
+        # limiting 230 bytes, so that a suffix (e.g. -annotations.json) can be added easily
+        limited_bytes = encoded[:max_length]
+        return limited_bytes.decode("utf-8", errors="replace")
```
This code does the same as errors="ignore" already does.

devvythelopper replied:
Never ever assume the AI is smart.

devvythelopper commented:
I can confirm that both the `--filename-length` option and the default filename trimming work. 

What is your policy on releasing new versions? When will this make it into an official release?

mkb79 commented:
@devvythelopper 
I'll have to document the changes in the README and CHANGELOG first. I'll think I'll release this at the weekend.

mkb79 commented:
@devvythelopper 
V0.3.3 is now live on PyPi. Thank you for your support!!!

devvythelopper commented:
Great, I'll be using this PRs codebase, I'll be following releases again once this appears in nixos.

mkb79 commented:
> Great, I'll be using this PRs codebase, I'll be following releases again once this appears in nixos.

@devvythelopper 
FYI: I've released v0.3.3 yesterday which contains the feature.
